massa-factory-worker/src/endorsement_factory.rs

```diff
@@ -0,0 +1,224 @@
+            if next_slot <= prev_slot {
+                next_slot = prev_slot
+                    .get_next_slot(self.cfg.thread_count)
+                    .expect("could not compute next slot");
+            }
+        }
+
+        // get the tiemstamp of the target slot
+        let next_instant = get_block_slot_timestamp(
+            self.cfg.thread_count,
+            self.cfg.t0,
+            self.cfg.genesis_timestamp,
+            next_slot,
+        )
+        .expect("could not get block slot timestamp")
+        .saturating_add(self.half_t0)
+        .estimate_instant(self.cfg.clock_compensation_millis)
+        .expect("could not estimate block slot instant");
+
+        (next_slot, next_instant)
+    }
+
+    /// Interruptibly wait until an instant or a stop signal
+    ///
+    /// # Return value
+    /// Returns `true` if the instant was reached, otherwise `false` if there was an interruption.
+    fn interruptible_wait_until(&self, duration: Instant) -> bool {
+        loop {
+            match self.factory_receiver.recv_deadline(duration) {
+                // message received => quit main loop
+                Ok(()) => return false,
+                // timeout => continue main loop
+                Err(mpsc::RecvTimeoutError::Timeout) => return true,
+                // channel disconnected (sender dropped) => quit main loop
+                Err(mpsc::RecvTimeoutError::Disconnected) => return false,
+            }
+        }
+    }
+
+    /// Process a slot: produce a block at that slot if one of the managed keys is drawn.
+    fn process_slot(&mut self, slot: Slot) {
+        // get endorsement producer addresses for that slot
+        let producer_addrs = match self.channels.selector.get_selection(slot) {
+            Ok(sel) => sel.endorsements,
+            Err(err) => {
+                warn!(
+                    "endorsement factory could not get selector draws for slot {}: {}",
+                    slot, err
+                );
+                return;
+            }
+        };
+
+        // get creators if they are managed by our wallet
+        let mut producers_indices: Vec<(KeyPair, usize)> = Vec::new();
+        {
+            let wallet = self.wallet.read().expect("could not lock wallet");
+            for (index, producer_addr) in producer_addrs.into_iter().enumerate() {
+                // check if the block producer address is handled by the wallet
+                let producer_keypair = match wallet.find_associated_keypair(&producer_addr) {
+                    // the selected block producer is managed locally => continue to attempt endorsement production
+                    Some(kp) => (kp, index),
+                    // the selected block producer is not managed locally => continue
+                    None => continue,
+                };
+            }
+        }
+
+        // quit if there is nothing to produce
+        if producers_indices.is_empty() {
+            return;
+        }
```

Comment:
`producers_indices` will always be empty here. Maybe it should be like the following proposal? I would also prefer a foreign method like `get_producers_indices`.

```suggestion
        // get creators if they are managed by our wallet
        let producers_indices: Vec<(KeyPair, usize)> = Vec::new();
        {
            let wallet = self.wallet.read().expect("could not lock wallet");
            for (index, producer_addr) in producer_addrs.into_iter().enumerate() {
                // check if the block producer address is handled by the wallet
                producers_indices.push(match wallet.find_associated_keypair(&producer_addr) {
                    // the selected block producer is managed locally => continue to attempt endorsement production
                    Some(kp) => (kp, index),
                    // the selected block producer is not managed locally => continue
                    None => continue,
                });
            }
        }

        // quit if there is nothing to produce
        if producers_indices.is_empty() {
            return;
        }
```